alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs

```diff
@@ -204,23 +252,92 @@ deriving via
     ) =>
     FromCBOR (Annotator (Tx era))
 
--- ===========================================
+-- ===============================================================
+-- From the specification, Figure 5 "Functions related to fees"
+-- ===============================================================
+
+isNonNativeScriptAddress :: forall era. ValidateScript era => Tx era -> Addr (Crypto era) -> Bool
+isNonNativeScriptAddress (TxConstr (Memo (TxRaw {_wits = w}) _)) addr =
+  case getValidatorHash addr of
+    Nothing -> False
+    Just hash ->
+      case Map.lookup hash (txscripts w) of
+        Nothing -> False
+        Just scr -> not (isNativeScript @era scr)
+
+feesOK ::
+  forall era.
+  ( UsesValue era,
+    AlonzoBody era,
+    UsesTxOut era,
+    ValidateScript era
+  ) =>
+  PParams era ->
+  Tx era ->
+  UTxO era ->
+  Bool
+feesOK pp tx (UTxO m) =
+  (bal >= txfee txb)
+    && (all (\txout -> not (isNonNativeScriptAddress tx (getField @"address" txout))) utxoFees)
+    && (minfee pp tx <= txfee txb)
+  where
+    txb = txbody tx
+    fees = inputs_fee txb
+    utxoFees = eval (fees ◁ m) -- compute the domain restriction to those inputs where fees are paid
+    bal = coin (balance @era (UTxO utxoFees))
+
+-- | The keys of all the inputs of the TxBody (both the inputs for fees, and the normal inputs).
+txins :: AlonzoBody era => TxBody era -> Set (TxIn (Crypto era))
+txins (TxBody {inputs = is, inputs_fee = fs}) = Set.union is fs
+
+txscriptfee :: Prices -> ExUnits -> Coin
+txscriptfee (Prices pr_mem pr_steps) (ExUnits mem steps) =
+  (mem <×> pr_mem) <+> (steps <×> pr_steps)
+
+-- TODO Check that txSize is the right functio-- Length of the serialised bytes
+minfee :: AlonzoBody era => PParams era -> Tx era -> Coin
+minfee pp tx =
+  ((txSize tx) <×> (a pp))
+    <+> (b pp)
+    <+> (txscriptfee (_prices pp) (exunits (txbody tx)))
+  where
+    a protparam = Coin (fromIntegral (_minfeeA protparam))
+    b protparam = Coin (fromIntegral (_minfeeB protparam))
+
+-- IT IS NOT AT ALL CLEAR what "validatorHash a" means in
+-- the specification. Maybe something like this?
```

Comment:
yes, this is correct. let's remove this comment before merging. we could add some prose to the spec if you think that that would help.

Reply:
done